rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+combined table.
+
+### Sampling the `joint_combiner`
+
+Currently, the `joint_combiner` is sampled using randomness that depends on the
+witness. However, where we have runtime or side-loaded tables, a malicious
+prover may be able to select values in those tables that absue knowledge of
+`joint_combiner` to create collisions.
+
+For example, to maliciously insert a value `(x, y, z)` for table with ID 1, the
+prover could select the value `x + y * joint_combiner + z * joint_combiner^2 +
+-1 * joint_combiner^max_joint_size` for an entry in table 2.
+
+Thus, we need to ensure that the `joint_combiner` depends on the values in
+runtime or side-loaded tables, so that the values in these tables cannot be
+chosen using knowledge of the `joint_combiner`. To do this, we must use the
+commitments to the values in these tables as part of the source of randomness
+for `joint_combiner` in the prover and verifier.
+
+In particular, this means that we must have a commitment per column for each
+type of table, so that the verifier can confirm the correct construction of the
+combined table.
+
+### Representing the combined fixed table
+
+The fixed table can be constructed at constraint system generation time, by
+extending all of the constituent tables to the same width, generating the
+appropriate table IDs array, and concatenating all of the tables. Concretely:
+
+Let `t[i][j][k]` be the `k`th element of the `j`th entry in the `i`th fixed
+table.
+Let `W` be the maximum width of all of the `t[i][j]`s.
+For any `t[i][j]` whose width `w` is less than `W`, pad it to width `W` by
+setting `t[i][j][k] = 0` for all `w < k <= W`.
+Form the combined table by concatenating all of these tables
+`FixedTable = t[0] || t[1] || t[2] || ...`
+and store in `TableID` the table ID that the corresponding row came from.
+Then, for every `x` and `y`, we have
+```
+FixedTable[x][k] = t[i][j][k]`
+TableID[x] = i
+where x = len(t[0]) + len(t[1]) + ... + len(t[i-1]) + j
+```
+
+This can be encoded as `W+1` polynomials in the constraint system, which we
+will reference as `FixedTable(i)` for `0 <= i < W` and `TableID`.
+
+### Representing the runtime table
+
+The runtime table can be considered as a 'mask' that we apply to the fixed
+table to introduce proof-specific data at runtime.
+
+We make the simplifying assumption that an entry in the runtime table has
+exactly 1 'fixed' entry, which can be used for indexing (or set to `0` if
+unused), and that the table has a fixed width of `X`. We can pad any narrower
+entries in the table to the full width `X` in the same way as the fixed tables
+above.
+
+We represent the masked values in the runtime table as `RuntimeTable(i)` for
+`1 <= i < X`.
+
+To specify whether a runtime table entry is applicable in the `i`th row, we use
+a selector polynomial `RuntimeTableSelector`. In order to reduce the polynomial
+degree of later checks, we add a constraint to check that the runtime entry is
+the zero vector wherever the selector is zero:
```

Comment:
an application of the Lagrange technique from Plonk?

Reply:
:man_shrugging: